lightning/impl/fw.py

```diff
@@ -0,0 +1,114 @@
+import numpy as np
+import scipy.sparse as sp
+
+from sklearn.base import BaseEstimator, RegressorMixin
+from sklearn.utils.extmath import safe_sparse_dot
+
+
+def _frank_wolfe(w_init, X, y, beta, max_iter=50, tol=1e-3, max_nz=None,
+                verbose=0):
+    """
+    Solve
+
+    0.5 * ||np.dot(X, w) - y||^2 s.t. ||w||_1 <= beta
+
+    by the Frank-Wolfe method.
+
+    The method can be seen as a greedy coordinate descent: it adds at most one
+    non-zero coefficient per iteration.
+    """
+    n_samples, n_features = X.shape
+
+    if sp.issparse(X):
+        X = X.tocsc()
+
+    w = w_init.copy()
+
+    for it in range(max_iter):
+        y_pred = safe_sparse_dot(X, w)
+        resid = beta * y_pred - y
+        neg_grad = -safe_sparse_dot(X.T, beta * resid)
+
+        atom = np.argmax(np.abs(neg_grad))
+        s = np.sign(neg_grad[atom])
+
+        error = np.dot(resid, resid)
+        dgap = s * neg_grad[atom] - np.dot(w, neg_grad)
+
+        if it == 0:
+            error_init = error
+            dgap_init = dgap
+
+        if verbose:
+            print "iter", it + 1
+            print "duality gap", dgap / dgap_init
+            print "error reduction", error / error_init
+            print "l1 norm", beta * np.sum(np.abs(w))
+            print "n_nz", np.sum(w != 0)
+            print
+
+        # Find optimal step size by exact line search.
+        Xs = s * X[:, atom]
+        if sp.issparse(Xs):
+            Xs_sq = np.dot(Xs.data, Xs.data)
+        else:
+            Xs_sq = np.dot(Xs, Xs)
+        y_pred_sq = np.dot(y_pred, y_pred)
+        b = (Xs - y_pred)
+        gamma = np.dot(resid, y_pred) - safe_sparse_dot(resid, Xs)
+        gamma /= beta * (Xs_sq - 2 * safe_sparse_dot(Xs.T, y_pred) + y_pred_sq)
```

Comment:
When running the example I get a warning at this line, because Xs - y_pred is zero. The line below fixes gamma, but I thought it might be worth it to point it out.